test/e2e/dr/restore-from-snapshot.go

```diff
@@ -0,0 +1,342 @@
+package dr
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+
+	"github.com/openshift/origin/test/e2e/upgrade"
+	exutil "github.com/openshift/origin/test/extended/util"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+)
+
+const (
+	sshOpts               = "-o StrictHostKeyChecking=no -o LogLevel=error -o ServerAliveInterval=30 -o ConnectionAttempts=100 -o ConnectTimeout=30"
+	proxyTemplate         = "ssh -A %s -W %%h:%%p core@%s 2>/dev/null"
+	scpTemplate           = "scp %s -o ProxyCommand=\"%s\" %s core@%s:%s"
+	sshTemplate           = "ssh %s -o ProxyCommand=\"%s\" core@%s \"%s\""
+	rollBackMachineConfig = "99-rollback-test"
+)
+
+var _ = g.Describe("[Feature:DisasterRecovery][Disruptive]", func() {
+	f := e2e.NewDefaultFramework("disaster-recovery")
+	f.SkipNamespaceCreation = true
+	f.SkipPrivilegedPSPBinding = true
+
+	oc := exutil.NewCLIWithoutNamespace("disaster-recovery")
+
+	g.It("Cluster should restore itself from etcd snapshot", func() {
+		config, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dynamicClient := dynamic.NewForConfigOrDie(config)
+		mcps := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "machineconfiguration.openshift.io",
+			Version:  "v1",
+			Resource: "machineconfigpools",
+		})
+		mc := dynamicClient.Resource(schema.GroupVersionResource{
+			Group:    "machineconfiguration.openshift.io",
+			Version:  "v1",
+			Resource: "machineconfigs",
+		})
+
+		bastionHost := setupSSHBastion(oc)
+		proxy := fmt.Sprintf(proxyTemplate, sshOpts, bastionHost)
+		defer removeSSHBastion(oc)
+
+		setMachineConfig("rollback-A.yaml", oc, mcps)
+
+		masters := getAllMasters(oc)
+		e2e.Logf("masters: %v", masters)
+		o.Expect(masters).NotTo(o.BeEmpty())
+		firstMaster := masters[0]
+		e2e.Logf("first master: %v", firstMaster)
+
+		e2e.Logf("Make etcd backup on first master")
+		runViaBastionSSH(firstMaster, proxy,
+			"sudo -i /bin/bash -x /usr/local/bin/etcd-snapshot-backup.sh /root/assets/backup/snapshot.db")
+		runViaBastionSSH(firstMaster, proxy,
+			"sudo -i install -o core -g core /root/assets/backup/snapshot.db /tmp/snapshot.db")
+		setMachineConfig("rollback-B.yaml", oc, mcps)
+
+		scpFileToHost(os.Getenv("KUBE_SSH_KEY_PATH"), proxy, "/home/core/.ssh/id_rsa", firstMaster)
```

Comment:
Discussed this with @deads2k - the code in lines L78-95 is running when the cluster is considered to be in a bad state already. 
This should be moved to a bash script, as it would be a great example for the customers of how DR is being tested in CI.